File: suites/osprey-compute-kernels/reference/cases/collatz/collatz.c
/* Collatz (3n+1) stopping time summed over 1..N — integer division (n/2) in deep recursion. */
#include <stdio.h>
#include <stdint.h>

static int64_t collatz(int64_t n) {
    if (n == 1) return 0;
    return (n % 2 == 0) ? 1 + collatz(n / 2) : 1 + collatz(3 * n + 1);
}

int main(void) {
    int64_t acc = 0;
    for (int64_t i = 1; i < 100001; i++) acc += collatz(i);
    printf("%lld\n", (long long)acc);
    return 0;
}

File: suites/osprey-compute-kernels/reference/cases/collatz/collatz.ml
(* Collatz (3n+1) stopping time summed over 1..N — integer division (n/2) in deep recursion. *)
let rec collatz n =
  if n = 1 then 0
  else if n mod 2 = 0 then 1 + collatz (n / 2)
  else 1 + collatz ((3 * n) + 1)

let () =
  let acc = ref 0 in
  for i = 1 to 100000 do
    acc := !acc + collatz i
  done;
  Printf.printf "%d\n" !acc
